### PR TITLE
fix: create volume when updating for the first time

### PR DESCRIPTION
**Pull request by icidasset**
Changes:

- **Fixed**: Volume id is not set on the account when first updating the volume.
- **Added**: `/ipfs/peers` endpoint with a list of IPFS node addresses.

### CODE REVIEW

File: fission-server/config/settings.toml [hunk 1]
```diff
@@ -2,18 +2,17 @@
...
-environment = "local"
-did = "did:web:localhost:3000"
-metrics_port = 4000
-port = 3000
-timeout_ms = 30000
+[ipfs]
+peers = [
+  "/ip4/127.0.0.1/tcp/4001",
+  "/ip4/127.0.0.1/tcp/4002/ws",
+  "/ip4/127.0.0.1/udp/4001/quic-v1/webtransport/certhash/uEiC4_sAMUfcxEJtqIlVWRGlHrTSSYuyk5Ulqfl6CjRiOHw/certhash/uEiANUHX9dRBqphQzZINo5WzkStJ7qevCr_2ZAUzLEbqoFw/p2p/12D3KooWLXH5BFzscChpdjtotAvv94hDR6bVxMKLHBHrSnugCRWs"
```
Comment (icidasset) [inline fission-server/config/settings.toml:14]: Normally you'd set this using an ENV variable, seeing that this last one will be different for each IPFS node. But keeping this here as an example of what to set it to.

File: fission-server/src/models/account.rs [hunk 2]
```diff
@@ -154,6 +155,13 @@ impl Account {
         conn: &mut Conn<'_>,
         cid: String,
     ) -> Result<NewVolumeRecord, diesel::result::Error> {
+        let ipfs = ipfs_api::IpfsClient::default();
+
+        if ipfs.pin_add(&cid, true).await.is_err() {
+            // FIXME: Use better error
+            return Err(diesel::result::Error::NotFound);
```
Comment (icidasset) [inline fission-server/src/models/account.rs:162]: This error is confusing can we do something better?

File: fission-server/src/routes/volume.rs [hunk 3]
```diff
@@ -64,7 +64,12 @@ pub async fn update_cid(
 ) -> AppResult<(StatusCode, Json<NewVolumeRecord>)> {
     let mut conn = db::connect(&state.db_pool).await?;
     let account = Account::find_by_username(&mut conn, username).await?;
-    let volume = account.update_volume_cid(&mut conn, &payload.cid).await?;
+
+    let volume: NewVolumeRecord = if account.volume_id.is_some() {
```
Comment (QuinnWilton) [inline fission-server/src/routes/volume.rs:68]: Does it make more sense to do this here, or to create the volume (possibly with a special genesis CID?) when the account is created?

File: fission-server/src/models/account.rs (hunk 2 above)
Comment (QuinnWilton) [reply fission-server/src/models/account.rs:162]: It looks like the error that comes back is described by [this](https://docs.rs/ipfs-api/latest/ipfs_api/enum.Error.html) enum, so if you wanted, you could define a `From` instance from that type to our error type, and differentiate between server and client errors.

File: fission-server/src/models/email_verification.rs [hunk 4]
```diff
@@ -71,6 +71,8 @@ impl VerificationCodeSender for EmailVerificationCodeSender {
     async fn send_code(&self, email: &str, code: &str) -> Result<()> {
         let message = self.message(email, code);
 
+        dbg!(&code);
```
Comment (QuinnWilton) [inline fission-server/src/models/email_verification.rs:74]: Leftover dbg! call

File: fission-server/src/routes/ipfs.rs [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+use crate::{
+    error::{AppError, AppResult},
+    settings::Settings,
+};
+
+use tracing::log;
+
+/// Render a list of IPFS node addresses
+pub async fn peers() -> AppResult<(StatusCode, Response)> {
+    let settings = Settings::load();
```
Comment (QuinnWilton) [inline fission-server/src/routes/ipfs.rs:19]: This is going to read the settings file on every request, and will also make testing more challenging by globally tying this piece of state to the filesystem.

It's possible that you might want to read this setting into the `AppState` when the application starts up, which will cut out that filesystem access and also give you a hook for configuring this option during tests.

File: fission-server/src/routes/volume.rs (hunk 3 above)
Comment (icidasset) [reply fission-server/src/routes/volume.rs:68]: For now we want to do it this way. But I think there were plans to create a WNFS and another key pair for it on the server side. We still need to flesh that out though, not sure if taking more power away from the client is a good idea 🤔

File: fission-server/src/routes/ipfs.rs (hunk 5 above)
Comment (icidasset) [reply fission-server/src/routes/ipfs.rs]: Oh good call! Done 👍

File: fission-server/src/models/account.rs (hunk 2 above)
Comment (icidasset) [reply fission-server/src/models/account.rs:162]: Yeah makes sense. This method is using the diesel error type though, I assume we want something different here then? `anyhow::Error`?

Comment (QuinnWilton) [reply fission-server/src/models/account.rs:162]: You'd potentially want to define your own error type and then define a translation from that to `AppError`, with that conversion happening in the route! You could also just return an `AppError` directly for now, though that does muddle the boundary between web and non-web concerns slightly.

**Review by QuinnWilton — APPROVED** [review]
This looks great! Thanks for making the settings changes :)

The errors can definitely benefit from some more thought, but I think that goes for the model errors in general, and I can imagine you all doing a more comprehensive pass through those at some point.

Anyhow is great, but one of its challenges is that we actually do want to, sometimes, inspect the specific errors being returned, so that we can map them more precisely to the correct HTTP status code. That means probably wanting a more carefully designed hierarchy of errors, and I think that's outside the scope of what you're doing here.

**Comment by QuinnWilton** [comment]
Oh! I did just notice that the tests are failing now though, because the `AppState` isn't being initialized with any peers during tests. Instead of making the peers required in the builder, I would recommend using `unwrap_or_default` to default to an empty list of peers if none are provided.

Alternatively, more idiomatic would potentially be to not use an `Option<Vec<_>>` there, and instead default to an empty `Vec`, and just have a `push_ipfs_peer` method for appending one to the end of the list. That would make it easier to incrementally build up the peers for tests with semi-shared logic.

**Comment by icidasset** [comment]
> Alternatively, more ergonomic

Sounds good, I'll implement this and make sure the tests pass 👍

**Pull request merged**